src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp

```diff
@@ -0,0 +1,387 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines which types are allowed, whether operations which certain types are
+/// allowed, and other type-specific properties and configuration for
+/// `TensorExpression`s
+///
+/// \details
+/// To add support for a data type, modify the templates in this file and the
+/// arithmetic operator overloads as necessary. Then, add tests as appropriate.
+
+#pragma once
+
+#include <cstddef>
+#include <limits>
+#include <type_traits>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/VectorImpl.hpp"
+
+namespace tenex {
+template <typename DataType>
+struct NumberAsExpression;
+
+namespace detail {
+/// \brief Whether or not `TensorExpression`s supports using a given type as a
+/// numeric term
+///
+/// \details
+/// To make it possible to use a new numeric data type as a term in
+/// `TensorExpression`s, add the type to this alias and adjust other templates
+/// in this file, as necessary.
+///
+/// \tparam X the arithmetic data type
+template <typename X>
+using is_supported_number_datatype =
+    std::bool_constant<std::is_same_v<X, double>>;
+
+/// \brief Whether or not `Tensor`s with the given data type are currently
+/// supported by `TensorExpression`s
+///
+/// \details
+/// To make it possible to use a new data type in a `Tensor` term in
+/// `TensorExpression`s, add the type to this alias and adjust other templates
+/// in this file, as necessary.
+///
+/// \tparam X the `Tensor` data type
+template <typename X>
+using is_supported_tensor_datatype =
+    std::bool_constant<std::is_same_v<X, double> or
+                       std::is_same_v<X, DataVector>>;
+
+/// \brief Whether or not the given type is a `VectorImpl` type
+///
+/// \tparam T the given type
+template <typename T>
+using is_vector = std::bool_constant<std::is_base_of_v<MarkAsVectorImpl, T>>;
```

Comment:
Definitely didn't see this defined there, so yes mine should be scrapped

Reply:
Wait sorry, mine differs in that it evaluates to `true` for both base `VectorImpl` and derived types, not just derived. So I think I will need to keep mine, if I understand `is_derived_of_vector` correctly

Reply:
In what case are you dealing with `VectorImpl` directly?

Reply:
Sorry, just removed my comment. Based on the alias' name, I had thought that `VectorImpl<double, DataVector>` would not be `true` because it isn't a _derived_ type of `VectorImpl`, but I just tested it and it's good. My `is_vector` is redundant, so I'll remove it.

Reply:
I'm dealing with `VectorImpl` directly in a few places where I wanted to extract the underlying value type and vector type, like here:

```
template <typename ValueType, typename VectorType>
struct lhs_datatype_is_assignable_to_rhs_datatype_impl<
    VectorImpl<std::complex<ValueType>, VectorType>, ValueType>
    : std::true_type {};
```

but I'm still reading through your comments, so maybe after I am done reading them, this won't be true anymore

Reply:
But in terms of when checking if something is a vector with `is_vector`, I don't think I'm actually using `VectorImpl`s directly, if that's what you're asking.

Regardless, my `is_vector` can be scrapped because it is equivalent to the existing one.